lib/dep-graph-builders/pnpm/parse-workspace.ts

```diff
@@ -0,0 +1,111 @@
+import { ScannedPnpmProject } from './types';
+
+const debug = debugModule('snyk-pnpm-workspaces');
+
+// Compute project versions map
+// This is needed because the lockfile doesn't present the version of
+// a project that's part of a workspace, we need to retrieve it from
+// its corresponding package.json
+function computeProjectVersionMaps(root: string, targets: string[]) {
+  const projectsVersionMap = {};
+  for (const target of targets) {
+    const directory = path.join(root, target);
+    const packageJsonFileName = path.join(directory, 'package.json');
+    const packageJson = getFileContents(root, packageJsonFileName);
+
+    try {
+      const parsedPkgJson = parsePkgJson(packageJson.content);
+      const projectVersion = parsedPkgJson.version;
+      projectsVersionMap[target] = projectVersion || 'undefined';
+    } catch (err: any) {
+      debug(
+        `Error getting version for project: ${packageJsonFileName}. ERROR: ${err}`,
+      );
+      continue;
+    }
+  }
+  return projectsVersionMap;
+}
+
+export const parsePnpmWorkspace = async (
+  root: string,
+  workspaceDir: string,
+  options: PnpmProjectParseOptions,
+) => {
+  const scannedProjects: ScannedPnpmProject[] = [];
+  const {
+    includeDevDeps,
+    includeOptionalDeps,
+    strictOutOfSync,
+    pruneWithinTopLevelDeps,
+  } = options;
+
+  const pnpmLockfileContents = getFileContents(
+    root,
+    path.join(workspaceDir, 'pnpm-lock.yaml'),
+  ).content;
+
+  const lockfileVersion = getPnpmLockfileVersion(pnpmLockfileContents);
+  const lockFileParser: PnpmLockfileParser = getPnpmLockfileParser(
+    pnpmLockfileContents,
+    lockfileVersion,
+  );
+
+  const projectVersionsMaps = computeProjectVersionMaps(
+    workspaceDir,
+    Object.keys(lockFileParser.importers),
+  );
+
+  for (const importer of Object.keys(lockFileParser.importers)) {
+    const resolvedImporterPath = path.join(workspaceDir, importer);
+    const pkgJsonFile = getFileContents(
+      root,
+      path.join(resolvedImporterPath, 'package.json'),
+    );
+
+    const pkgJson: PackageJsonBase = parsePkgJson(pkgJsonFile.content);
+
+    lockFileParser.workspaceArgs = {
+      isWorkspace: true,
+      projectsVersionMap: projectVersionsMaps,
+    };
+
+    try {
+      const depGraph: DepGraph = await buildDepGraphPnpm(
+        lockFileParser,
+        pkgJson,
+        {
+          includeDevDeps,
+          strictOutOfSync,
+          includeOptionalDeps,
+          pruneWithinTopLevelDeps,
+        },
+        importer,
+      );
+
+      const project: ScannedPnpmProject = {
+        packageManager: 'pnpm',
+        targetFile: path.relative(root, pkgJsonFile.fileName),
+        depGraph,
+        plugin: {
+          name: 'snyk-nodejs-lockfile-parser',
+          runtime: process.version,
+        },
+      };
+      scannedProjects.push(project);
+    } catch (e) {
+      debug(`Error process workspace: ${pkgJsonFile.fileName}. ERROR: ${e}`);
```

Comment:
shouldn't we throw an actual error here?

Reply:
No, this is the flow for '--all-projects', so we shouldn't interrupt the processing of the other projects.